File: apps/family-service/src/modules/goal/goal.controller.ts
import { Body, Controller, Delete, Get, Param, Patch, Post, Query } from '@nestjs/common';
import { ApiOperation, ApiParam, ApiResponse, ApiTags } from '@nestjs/swagger';
import { GoalService } from './goal.service';
import {
  CreateGoalInput,
  Goal,
  GoalConnection,
  GoalListInput,
  UpdateGoalProgressInput,
} from './goal.types';

@ApiTags('goals')
@Controller('api/v1/goals')
export class GoalController {
  constructor(private readonly goalService: GoalService) {}

  @Get()
  @ApiOperation({ summary: 'List goals' })
  @ApiResponse({ status: 200, description: 'Paginated goal collection.', type: GoalConnection })
  listGoals(@Query() query: GoalListInput): Promise<GoalConnection> {
    return this.goalService.listGoals(query);
  }

  @Get(':id')
  @ApiOperation({ summary: 'Get a goal by ID' })
  @ApiParam({ name: 'id', description: 'Goal identifier' })
  @ApiResponse({ status: 200, description: 'Goal record.', type: Goal })
  @ApiResponse({ status: 404, description: 'Goal not found.' })
  getGoal(@Param('id') id: string): Promise<Goal | null> {
    return this.goalService.getGoal(id);
  }

  @Post()
  @ApiOperation({ summary: 'Create a goal' })
  @ApiResponse({ status: 201, description: 'Created goal.', type: Goal })
  createGoal(@Body() input: CreateGoalInput): Promise<Goal> {
    return this.goalService.createGoal(input);
  }

  @Patch(':id/progress')
  @ApiOperation({ summary: 'Update goal progress' })
  @ApiParam({ name: 'id', description: 'Goal identifier' })
  @ApiResponse({ status: 200, description: 'Updated goal.', type: Goal })
  @ApiResponse({ status: 404, description: 'Goal not found.' })
  updateGoalProgress(
    @Param('id') id: string,
    @Body() input: UpdateGoalProgressInput,
  ): Promise<Goal | null> {
    return this.goalService.updateGoalProgress(id, input);
  }

  @Delete(':id')
  @ApiOperation({ summary: 'Delete a goal' })
  @ApiParam({ name: 'id', description: 'Goal identifier' })
  @ApiResponse({ status: 200, description: 'Whether deletion succeeded.', type: Boolean })
  deleteGoal(@Param('id') id: string): Promise<boolean> {
    return this.goalService.deleteGoal(id);
  }
}

File: apps/identity-service/src/main/java/com/family/goal/controller/GoalController.java
package com.family.goal.controller;

import com.family.goal.dto.CreateGoalRequest;
import com.family.goal.dto.GoalListResponse;
import com.family.goal.dto.GoalResponse;
import com.family.goal.dto.UpdateGoalProgressRequest;
import com.family.goal.service.GoalService;
import io.swagger.v3.oas.annotations.Operation;
import io.swagger.v3.oas.annotations.responses.ApiResponse;
import io.swagger.v3.oas.annotations.responses.ApiResponses;
import io.swagger.v3.oas.annotations.security.SecurityRequirement;
import io.swagger.v3.oas.annotations.tags.Tag;
import jakarta.validation.Valid;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;

import java.security.Principal;
import java.util.Map;
import java.util.UUID;

@Tag(name = "Goals", description = "Goal management")
@RestController
@RequestMapping("/api/v1/goals")
public class GoalController {

    private final GoalService goalService;

    public GoalController(GoalService goalService) {
        this.goalService = goalService;
    }

    private UUID getUserId(Principal principal) {
        if (principal == null) {
            throw new IllegalStateException("Authentication required");
        }
        return UUID.fromString(principal.getName());
    }

    @Operation(summary = "List goals", security = @SecurityRequirement(name = "bearerAuth"))
    @ApiResponse(responseCode = "200", description = "List of goals")
    @GetMapping
    public ResponseEntity<GoalListResponse> findAll(
            @RequestParam(required = false) UUID memberId,
            @RequestParam(required = false) String type,
            @RequestParam(required = false) String status,
            @RequestParam(required = false) Integer first,
            @RequestParam(required = false) String after,
            Principal principal) {
        return ResponseEntity.ok(goalService.findAll(getUserId(principal), memberId, type, status, first, after));
    }

    @Operation(summary = "Get goal by ID", security = @SecurityRequirement(name = "bearerAuth"))
    @ApiResponses({
            @ApiResponse(responseCode = "200", description = "Goal found"),
            @ApiResponse(responseCode = "404", description = "Goal not found")
    })
    @GetMapping("/{id}")
    public ResponseEntity<GoalResponse> findById(@PathVariable UUID id, Principal principal) {
        return ResponseEntity.ok(goalService.findById(getUserId(principal), id));
    }

    @Operation(summary = "Create a goal", security = @SecurityRequirement(name = "bearerAuth"))
    @ApiResponses({
            @ApiResponse(responseCode = "201", description = "Goal created"),
            @ApiResponse(responseCode = "400", description = "Validation failed")
    })
    @PostMapping
    public ResponseEntity<GoalResponse> create(@Valid @RequestBody CreateGoalRequest request, Principal principal) {
        GoalResponse goal = goalService.create(getUserId(principal), request);
        return ResponseEntity.status(HttpStatus.CREATED).body(goal);
    }

    @Operation(summary = "Update goal progress", security = @SecurityRequirement(name = "bearerAuth"))
    @ApiResponses({
            @ApiResponse(responseCode = "200", description = "Goal updated"),
            @ApiResponse(responseCode = "404", description = "Goal not found")
    })
    @PatchMapping("/{id}/progress")
    public ResponseEntity<GoalResponse> updateProgress(@PathVariable UUID id,
                                                       @Valid @RequestBody UpdateGoalProgressRequest request,
                                                       Principal principal) {
        return ResponseEntity.ok(goalService.updateProgress(getUserId(principal), id, request));
    }

    @Operation(summary = "Delete a goal", security = @SecurityRequirement(name = "bearerAuth"))
    @ApiResponses({
            @ApiResponse(responseCode = "200", description = "Goal deleted"),
            @ApiResponse(responseCode = "404", description = "Goal not found")
    })
    @DeleteMapping("/{id}")
    public ResponseEntity<Map<String, String>> delete(@PathVariable UUID id, Principal principal) {
        goalService.delete(getUserId(principal), id);
        return ResponseEntity.ok(Map.of("message", "Goal deleted"));
    }

    @Operation(summary = "Expire overdue goals", security = @SecurityRequirement(name = "bearerAuth"))
    @PostMapping("/expire-due")
    public ResponseEntity<Map<String, Integer>> expireDueGoals() {
        int expired = goalService.expireDueGoals();
        return ResponseEntity.ok(Map.of("expired", expired));
    }
}
